node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -0,0 +1,199 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::channel::oneshot;
+use std::collections::{HashMap, VecDeque};
+
+use polkadot_node_network_protocol::PeerId;
+use polkadot_node_primitives::PoV;
+use polkadot_primitives::v2::{CandidateHash, CandidateReceipt, CollatorId, Hash, Id as ParaId};
+
+use crate::{ProspectiveParachainsMode, LOG_TARGET, MAX_CANDIDATE_DEPTH};
+
+/// Candidate hash paired with the parent head hash.
+#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
+pub struct ProspectiveCandidate(pub Option<(CandidateHash, Hash)>);
+
+impl ProspectiveCandidate {
+	pub fn candidate_hash(&self) -> Option<CandidateHash> {
+		self.0.as_ref().map(|c| c.0)
+	}
+}
+
+/// Identifier of a fetched collation.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct FetchedCollation {
+	pub relay_parent: Hash,
+	pub para_id: ParaId,
+	pub candidate_hash: CandidateHash,
+	pub collator_id: CollatorId,
+}
+
+impl From<&CandidateReceipt<Hash>> for FetchedCollation {
+	fn from(receipt: &CandidateReceipt<Hash>) -> Self {
+		let descriptor = receipt.descriptor();
+		Self {
+			relay_parent: descriptor.relay_parent,
+			para_id: descriptor.para_id,
+			candidate_hash: receipt.hash(),
+			collator_id: descriptor.collator.clone(),
+		}
+	}
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct PendingCollation {
+	pub relay_parent: Hash,
+	pub para_id: ParaId,
+	pub peer_id: PeerId,
+	pub prospective_candidate: ProspectiveCandidate,
+	pub commitments_hash: Option<Hash>,
+}
+
+impl PendingCollation {
+	pub fn new(
+		relay_parent: Hash,
+		para_id: ParaId,
+		peer_id: &PeerId,
+		prospective_candidate: ProspectiveCandidate,
+	) -> Self {
+		Self {
+			relay_parent,
+			para_id,
+			peer_id: peer_id.clone(),
+			prospective_candidate,
+			commitments_hash: None,
+		}
+	}
+}
+
+pub type CollationEvent = (CollatorId, PendingCollation);
+
+pub type PendingCollationFetch =
+	(CollationEvent, std::result::Result<(CandidateReceipt, PoV), oneshot::Canceled>);
+
+/// The status of the collations in [`CollationsPerRelayParent`].
+#[derive(Debug, Clone, Copy)]
+pub enum CollationStatus {
+	/// We are waiting for a collation to be advertised to us.
+	Waiting,
+	/// We are currently fetching a collation.
+	Fetching,
+	/// We are waiting that a collation is being validated.
+	WaitingOnValidation,
+	/// We have seconded a collation.
+	Seconded,
+}
+
+impl Default for CollationStatus {
+	fn default() -> Self {
+		Self::Waiting
+	}
+}
+
+impl CollationStatus {
+	/// Downgrades to `Waiting`, but only if `self != Seconded`.
+	fn back_to_waiting(&mut self, relay_parent_mode: ProspectiveParachainsMode) {
+		match self {
+			Self::Seconded =>
+				if relay_parent_mode.is_enabled() {
+					// With async backing enabled it's allowed to
+					// second more candidates.
+					*self = Self::Waiting
```

Comment:
It can happen that we try to dequeue collation, but then some check fails and we don't start fetching
It's better to reset it unconditionally to avoid queuing more collations into memory